src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -50,6 +51,24 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
         private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource> source, int size)
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
+
+            // TODO what should the threshold be?
+            if (size > 1000)
```

Comment:
This just handles the first chunk, though. If this were to have its own full iterator we'd need to copy over most of `ChunkIterator`.

Reply:
That approach solves the OOM problem, but it doesn't prevent the second chunk from needlessly allocating a new array of length ~`MaxArrayLength`.

Reply:
You're correct. My thought is that this matters (proportionally) less with each full chunk.

If we create an array of length `size` for the first chunk when we only *needed* to allocate an array of length X, then the "waste factor" is size / X which can be very large.

In contrast, if we have more than one chunk and the waste only occurs on the last chunk, then the waste factor is at most 2.

Reply:
Here's a quick benchmark comparing 3 modes:
* "Native" (the current implementation)
* "List" (reuse a `List<T>` instance to build each chunk)
* "First" (use `List<T>` to build the first chunk *regardless of chunk size*)

In these scenarios, List and First are both **much** better in the lopsided case (1000,1000000) which is expected. In the worst case for First (1000,999), List is slightly better. In other cases, First comes in slightly/moderately ahead of List. This largely lines up with what intuition suggests. 

As you say, First is a bit more complex than list (~10 extra lines of code) but on balance has slightly better performance. I'm not sure whether it is enough to be worth it. Thoughts?

|      Method |        Source Count,Chunk Size |      Mean |     Error |     StdDev |    Median | Ratio | RatioSD |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|------------ |------------- |----------:|----------:|-----------:|----------:|------:|--------:|---------:|---------:|---------:|----------:|
| **ChunkNative** |     **1000,100** |  **21.74 μs** |  **0.313 μs** |   **0.293 μs** |  **21.70 μs** |  **1.00** |    **0.00** |   **2.6550** |        **-** |        **-** |      **8 KB** |
|   ChunkList |     1000,100 |  24.92 μs |  0.470 μs |   0.440 μs |  24.80 μs |  1.15 |    0.03 |   3.3569 |        - |        - |     10 KB |
|  ChunkFirst |     1000,100 |  23.31 μs |  0.461 μs |   1.050 μs |  23.17 μs |  1.08 |    0.04 |   3.3569 |        - |        - |     10 KB |
|             |              |           |           |            |           |       |         |          |          |          |           |
| **ChunkNative** | **1000,1000000** | **883.51 μs** | **99.235 μs** | **291.040 μs** | **787.80 μs** |  **1.00** |    **0.00** | **173.8281** | **170.8984** | **170.8984** |  **7,821 KB** |
|   ChunkList | 1000,1000000 |  25.58 μs |  0.480 μs |   0.426 μs |  25.47 μs |  0.04 |    0.00 |   7.8735 |        - |        - |     24 KB |
|  ChunkFirst | 1000,1000000 |  25.54 μs |  0.497 μs |   0.440 μs |  25.59 μs |  0.04 |    0.00 |   7.8735 |        - |        - |     24 KB |
|             |              |           |           |            |           |       |         |          |          |          |           |
| **ChunkNative** |     **1000,999** |  **22.88 μs** |  **0.439 μs** |   **1.078 μs** |  **22.52 μs** |  **1.00** |    **0.00** |   **5.1270** |        **-** |        **-** |     **16 KB** |
|   ChunkList |     1000,999 |  27.74 μs |  0.552 μs |   1.090 μs |  27.64 μs |  1.20 |    0.07 |   7.8735 |        - |        - |     24 KB |
|  ChunkFirst |     1000,999 |  29.22 μs |  0.805 μs |   2.360 μs |  29.09 μs |  1.27 |    0.14 |  10.4370 |        - |        - |     32 KB |
|             |              |           |           |            |           |       |         |          |          |          |           |
| **ChunkNative** |    **10000,999** | **231.38 μs** |  **4.590 μs** |   **6.127 μs** | **232.73 μs** |  **1.00** |    **0.00** |  **27.8320** |        **-** |        **-** |     **86 KB** |
|   ChunkList |    10000,999 | 267.46 μs |  6.508 μs |  18.776 μs | 264.60 μs |  1.12 |    0.06 |  30.7617 |        - |        - |     95 KB |
|  ChunkFirst |    10000,999 | 221.95 μs |  4.243 μs |   4.167 μs | 222.43 μs |  0.96 |    0.03 |  33.2031 |        - |        - |    103 KB |

Reply:
Could you share the benchmark code? What element type did you use for the benchmarks? The shared List approach might be slightly disadvantaged for reference types since [`List.Clear()`  is more expensive](https://github.com/dotnet/runtime/blob/c0db07b3f3f093b5f4cd27e1f5e8aa54adad049d/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L292).

If the numbers above are representative, I would say that the shared list approach is probably good enough.

Reply:
@eiriktsarpalis here's the benchmark code https://gist.github.com/madelson/ec76bc142b6c015a38ed077e80eda3d0

I was using `string` as the element type for exactly the reason you describe: it captures the full cost of `Clear()`.

Reply:
In that case, I'd just go for a simple list-based implementation.